crates/bevy_ecs/src/reflect.rs

```diff
@@ -406,26 +406,27 @@ impl_from_reflect_value!(Entity);
 
 #[derive(Clone)]
 pub struct ReflectMapEntities {
-    map_entities: fn(&mut World, &EntityMap) -> Result<(), MapEntitiesError>,
+    map_entities: fn(&mut World, &mut EntityMapper) -> Result<(), MapEntitiesError>,
 }
 
 impl ReflectMapEntities {
     pub fn map_entities(
         &self,
         world: &mut World,
-        entity_map: &EntityMap,
+        entity_map: &mut EntityMap,
     ) -> Result<(), MapEntitiesError> {
-        (self.map_entities)(world, entity_map)
+        entity_map.with_mapper(world, self.map_entities)
     }
 }
 
 impl<C: Component + MapEntities> FromType<C> for ReflectMapEntities {
     fn from_type() -> Self {
         ReflectMapEntities {
-            map_entities: |world, entity_map| {
-                for entity in entity_map.values() {
+            map_entities: |world, entity_mapper| {
+                let entities = entity_mapper.get_map().values().collect::<Vec<Entity>>();
+                for entity in entities.iter().cloned() {
                     if let Some(mut component) = world.get_mut::<C>(entity) {
-                        component.map_entities(entity_map)?;
+                        component.map_entities(entity_mapper)?;
```

Comment:
If I understand you right, what you're really trying to get at is if the closure should be specialized to return a result so that the rest of `with_mapper` can react to it. I think no, in that saving the mapper should really always happen even if there was an error somewhere in mapping. Saving the mapper just actually allocates in the world the ids that it handed out in mapping, and this should always happen.

When mapping fails for whatever reason we're left in a bad state, with things only partially mapped and possibly pointing at entities that already exist in the destination world. What we might consider doing, though, is just making it so that mapping isn't allowed to fail. With `get_or_alloc` now existing I can't think of a good reason to ever call `get` or ever to return an error from a `MapEntities` implementation, but I wasn't sure if there was some weird use case I was missing and didn't want to unnecessarily expand the blast radius. That would be reasonable to just go ahead and do though.